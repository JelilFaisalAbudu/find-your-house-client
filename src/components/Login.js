/* eslint-disable no-unused-vars */
/* eslint-disable react/button-has-type */
/* eslint-disable jsx-a11y/label-has-associated-control */
/* eslint-disable no-underscore-dangle */
/* eslint-disable react/prop-types */
import React, { useEffect, useState, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect } from 'react-router-dom';

/* eslint-disable consistent-return */
import Form from 'react-validation/build/form';
import Input from 'react-validation/build/input';
import CheckButton from 'react-validation/build/button';

import { login } from '../redux/actions/auth';
import history from '../helpers/history';
import { spyNoScroll } from '../helpers/dom';

const required = value => {
  if (!value) {
    return (
      <div className="alert alert-danger" role="alert">
        This field is required!
      </div>
    );
  }
};

const Login = () => {
  const form = useRef();
  const checkBtn = useRef();

  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [loading, setLoading] = useState(false);

  const { isLoggedIn } = useSelector(state => state.auth);
  const { message } = useSelector(state => state.message);

  const dispatch = useDispatch();

  const onChangeEmail = e => {
    const email = e.target.value;
    setEmail(email);
  };

  const onChangePassword = e => {
    const password = e.target.value;
    setPassword(password);
  };

  const handleLogin = e => {
    e.preventDefault();

    setLoading(true);

    form.current.validateAll();

    if (checkBtn.current.context._errors.length === 0) {
      dispatch(login(email, password))
        .then(() => {
          history.push('/home');
        })
        .catch(() => {
          setLoading(false);
        });
    } else {
      setLoading(false);
    }
  };

  if (isLoggedIn) {
    return <Redirect to="/home" />;
  }

  spyNoScroll();

  return (
    <div className="col-md-12">
      <div className="card card-container">
        <img
          src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
          alt="profile-img"
          className="profile-img-card"
        />

        <Form onSubmit={handleLogin} ref={form}>
          <div className="form-group">
            <label htmlFor="email">Email</label>
            <Input
              type="email"
              className="form-control"
              name="email"
              value={email}
              onChange={onChangeEmail}
              validations={[required]}
            />
          </div>

          <div className="form-group">
            <label htmlFor="password">Password</label>
            <Input
              type="password"
              className="form-control"
              name="password"
              value={password}
              onChange={onChangePassword}
              validations={[required]}
            />
          </div>

          <div className="form-group">
            <button className="btn btn-primary btn-block btn-submit" disabled={loading}>
              {loading && (
                <span className="spinner-border spinner-border-sm" />
              )}
              <span className="login-text">Login</span>
            </button>
          </div>

          {message && (
            <div className="form-group">
              <div className="alert alert-danger" role="alert">
                {message}
              </div>
            </div>
          )}
          <CheckButton style={{ display: 'none' }} ref={checkBtn} />
        </Form>
      </div>
    </div>
  );
};

export default Login;
