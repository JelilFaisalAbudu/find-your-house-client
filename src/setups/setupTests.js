import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-17-updated';

Enzyme.configure({ adapter: new Adapter() });
