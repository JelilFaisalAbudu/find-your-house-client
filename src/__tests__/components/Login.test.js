import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import '../../setups/setupTests';
import Login from '../../components/Login';
import store from '../../redux/store';

const setup = () => mount(
  <Provider store={store}>
    <Login />
  </Provider>,
);

describe('Login', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = setup();
  });

  test('renders without crashing', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('renders the login form correctly', async () => {
    const button = await wrapper.find('button.btn-login');
    expect(button.length).toEqual(1);
  });
});
