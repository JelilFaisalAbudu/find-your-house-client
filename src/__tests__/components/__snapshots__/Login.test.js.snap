// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login renders without crashing 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <Login>
    <div
      className="col-md-12"
    >
      <div
        className="card card-container"
      >
        <img
          alt="profile-img"
          className="profile-img-card"
          src="//ssl.gstatic.com/accounts/ui/avatar_2x.png"
        />
        <Form(t)
          onSubmit={[Function]}
        >
          <t
            getValues={[Function]}
            hideError={[Function]}
            onSubmit={[Function]}
            showError={[Function]}
            validate={[Function]}
            validateAll={[Function]}
          >
            <form
              onSubmit={[Function]}
            >
              <div
                className="form-group"
              >
                <label
                  htmlFor="email"
                >
                  Email
                </label>
                <Control(s)
                  className="form-control"
                  name="email"
                  onChange={[Function]}
                  type="email"
                  validations={
                    Array [
                      [Function],
                    ]
                  }
                  value=""
                >
                  <s
                    className="form-control"
                    error={
                      <div
                        className="alert alert-danger"
                        role="alert"
                      >
                        This field is required!
                      </div>
                    }
                    name="email"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="email"
                    value=""
                  >
                    <div>
                      <input
                        className="form-control"
                        name="email"
                        onBlur={[Function]}
                        onChange={[Function]}
                        type="email"
                        value=""
                      />
                    </div>
                  </s>
                </Control(s)>
              </div>
              <div
                className="form-group"
              >
                <label
                  htmlFor="password"
                >
                  Password
                </label>
                <Control(s)
                  className="form-control"
                  name="password"
                  onChange={[Function]}
                  type="password"
                  validations={
                    Array [
                      [Function],
                    ]
                  }
                  value=""
                >
                  <s
                    className="form-control"
                    error={
                      <div
                        className="alert alert-danger"
                        role="alert"
                      >
                        This field is required!
                      </div>
                    }
                    name="password"
                    onBlur={[Function]}
                    onChange={[Function]}
                    type="password"
                    value=""
                  >
                    <div>
                      <input
                        className="form-control"
                        name="password"
                        onBlur={[Function]}
                        onChange={[Function]}
                        type="password"
                        value=""
                      />
                    </div>
                  </s>
                </Control(s)>
              </div>
              <div
                className="form-group"
              >
                <button
                  className="btn btn-primary btn-block btn-submit"
                  disabled={false}
                >
                  <span
                    className="login-text"
                  >
                    Login
                  </span>
                </button>
              </div>
              <Button(s)
                style={
                  Object {
                    "display": "none",
                  }
                }
              >
                <s
                  hasErrors={true}
                  style={
                    Object {
                      "display": "none",
                    }
                  }
                >
                  <button
                    disabled={true}
                    style={
                      Object {
                        "display": "none",
                      }
                    }
                  />
                </s>
              </Button(s)>
            </form>
          </t>
        </Form(t)>
      </div>
    </div>
  </Login>
</Provider>
`;
